drivers/storage/efs/storage/efs_storage.go

```diff
@@ -50,37 +61,210 @@ func (d *driver) Name() string {
-			&credentials.EnvProvider{},
-			&credentials.SharedCredentialsProvider{},
-			&ec2rolecreds.EC2RoleProvider{
-				Client: ec2metadata.New(session.New()),
+	writeHkey(hkey, region)
+	writeHkey(hkey, endpoint)
+	writeHkey(hkey, &akey)
+	ckey = fmt.Sprintf("%x", hkey.Sum(nil))
+
+	// if the session is cached then return it
+	if svc, ok := sessions[ckey]; ok {
+		ctx.WithField(cacheKeyC, ckey).Debug("using cached efs service")
+		return svc, nil
+	}
+
+	var (
+		skey   = d.secretKey()
+		fields = map[string]interface{}{
+			efs.AccessKey: akey,
+			efs.Tag:       d.tag,
+			cacheKeyC:     ckey,
+		}
+	)
+
+	if skey == "" {
+		fields[efs.SecretKey] = ""
+	} else {
+		fields[efs.SecretKey] = "******"
+	}
+	if region != nil {
+		fields[efs.Region] = *region
+	}
+	if endpoint != nil {
+		fields[efs.Endpoint] = *endpoint
+	}
+	if endpointEC2 != nil {
+		fields[efs.EndpointEC2] = *endpointEC2
+	}
+
+	ctx.WithFields(fields).Debug("efs service connetion attempt")
+	sess := session.New()
+
+	awsLogLevel := aws.LogOff
+	if ll, ok := context.GetLogLevel(ctx); ok {
+		switch ll {
+		case log.DebugLevel:
+			awsLogLevel = aws.LogDebugWithHTTPBody
+		case log.InfoLevel:
+			awsLogLevel = aws.LogDebug
+		}
+	}
+
+	configEFS := &aws.Config{
+		Region:     region,
+		Endpoint:   endpoint,
+		MaxRetries: d.maxRetries,
+		Credentials: credentials.NewChainCredentials(
+			[]credentials.Provider{
+				&credentials.StaticProvider{
+					Value: credentials.Value{
+						AccessKeyID:     akey,
+						SecretAccessKey: skey,
+					},
+				},
+				&credentials.EnvProvider{},
+				&credentials.SharedCredentialsProvider{},
+				&ec2rolecreds.EC2RoleProvider{
+					Client: ec2metadata.New(sess),
+				},
 			},
-		})
+		),
+		Logger:   &awsLogger{ctx},
+		LogLevel: aws.LogLevel(awsLogLevel),
+	}
+
+	configEC2 := &aws.Config{
+		Region:     region,
+		Endpoint:   endpointEC2,
+		MaxRetries: d.maxRetries,
+		Credentials: credentials.NewChainCredentials(
+			[]credentials.Provider{
+				&credentials.StaticProvider{
+					Value: credentials.Value{
+						AccessKeyID:     akey,
+						SecretAccessKey: skey,
+					},
+				},
+				&credentials.EnvProvider{},
+				&credentials.SharedCredentialsProvider{},
+				&ec2rolecreds.EC2RoleProvider{
+					Client: ec2metadata.New(sess),
+				},
+			},
+		),
+		Logger:   &awsLogger{ctx},
+		LogLevel: aws.LogLevel(awsLogLevel),
+	}
 
-	ctx.WithFields(fields).Info("storage driver initialized")
+	svc := &awsService{
+		awsec2.New(sess, configEC2),
+		awsefs.New(sess, configEFS),
+	}
+	sessions[ckey] = svc
+
+	ctx.WithFields(fields).Info("efs service connetion created & cached")
+
+	return svc, nil
+}
+
+type awsService struct {
+	ec2 *awsec2.EC2
+	efs *awsefs.EFS
+}
+
+type awsLogger struct {
+	ctx types.Context
+}
+
+func (a *awsLogger) Log(args ...interface{}) {
+	a.ctx.Infoln(args...)
+}
+
+func mustSession(ctx types.Context) *awsService {
+	return context.MustSession(ctx).(*awsService)
+}
+
+func mustInstanceIDID(ctx types.Context) *string {
```

Comment:
Hi @mhrabovcin,

Nope. It returns the address of the instance ID's `ID` field or panics if the instance ID is not found in the context.